packages/sdk/src/cosmwasmclient.ts

```diff
@@ -0,0 +1,228 @@
+import { Encoding } from "@iov/encoding";
+
+import { makeSignBytes, marshalTx } from "./encoding";
+import { findAttribute, Log, parseLogs } from "./logs";
+import { RestClient } from "./restclient";
+import {
+  Coin,
+  MsgExecuteContract,
+  MsgInstantiateContract,
+  MsgStoreCode,
+  StdFee,
+  StdSignature,
+} from "./types";
+
+const defaultUploadFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "1000000", // one million
+};
+
+const defaultInitFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "500000", // 500k
+};
+
+const defaultExecFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "200000", // 200k
+};
+
+export interface SigningCallback {
+  (signBytes: Uint8Array): Promise<StdSignature>;
+}
+
+interface SigningData {
+  readonly senderAddress: string;
+  readonly signCallback: SigningCallback;
+}
+
+export interface GetNonceResult {
+  readonly accountNumber: number;
+  readonly sequence: number;
+}
+
+export interface PostTxResult {
+  readonly logs: readonly Log[];
+  readonly rawLog: string;
+  readonly transactionHash: string;
+}
+
+export class CosmWasmClient {
+  public static makeReadOnly(url: string): CosmWasmClient {
+    return new CosmWasmClient(url);
+  }
+
+  public static makeWritable(
+    url: string,
+    senderAddress: string,
+    signCallback: SigningCallback,
+  ): CosmWasmClient {
+    return new CosmWasmClient(url, {
+      senderAddress: senderAddress,
+      signCallback: signCallback,
+    });
+  }
+
+  private readonly restClient: RestClient;
+  private readonly signingData: SigningData | undefined;
+
+  private get senderAddress(): string {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.senderAddress;
+  }
+
+  private get signCallback(): SigningCallback {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.signCallback;
+  }
+
+  private constructor(url: string, signingData?: SigningData) {
+    this.restClient = new RestClient(url);
+    this.signingData = signingData;
+  }
+
+  public async chainId(): Promise<string> {
+    const response = await this.restClient.nodeInfo();
```

Comment:
We can cache this on first call, it is still a Promise, but one less rest call when we use the client for multiple txs

Reply:
I like how you think :)